test/include/psen_scan_v2/message_validator.h

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) 2020 Pilz GmbH & Co. KG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#ifndef PSEN_SCAN_V2_MESSAGE_COLLECTOR_H
+#define PSEN_SCAN_V2_MESSAGE_COLLECTOR_H
+
+#include <future>
+#include <map>
+#include <vector>
+#include "psen_scan_v2/angle_conversions.h"
+#include "psen_scan_v2/dist.h"
+
+#include <ros/ros.h>
+
+namespace psen_scan_v2_test
+{
+void addScanToBin(const sensor_msgs::LaserScanConstPtr& scan, std::map<int16_t, NormalDist>& bin)
+{
+  if (scan == nullptr)
+  {
+    throw;
+  }
+
+  for (size_t i = 0; i < scan->ranges.size(); ++i)
+  {
+    auto bin_addr = psen_scan_v2::radToTenthDegree(scan->angle_min + scan->angle_increment * i);
+
+    if (bin.find(bin_addr) == bin.end())
+    {
+      bin.emplace(bin_addr, NormalDist{});
+      // Create bin
+    }
+    bin[bin_addr].update(scan->ranges[i]);
+  }
+}
+
+std::map<int16_t, NormalDist> binsFromScans(std::vector<sensor_msgs::LaserScanConstPtr> scans)
+{
+  std::map<int16_t, NormalDist> bins;
+  std::for_each(
+      scans.cbegin(), scans.cend(), [&bins](const sensor_msgs::LaserScanConstPtr& scan) { addScanToBin(scan, bins); });
+  return bins;
+}
+
+template <typename MsgType>
+class MessageValidator
+{
+public:
+  MessageValidator(ros::NodeHandle& nh, std::map<int16_t, NormalDist> bins_expected)
+    : nh_(nh), bins_expected_(bins_expected){};
+
+  typedef boost::shared_ptr<MsgType const> MsgTypeConstPtr;
+
+  void scanCb(const MsgTypeConstPtr scan, size_t n_msgs)
+  {
+    ROS_INFO_THROTTLE(5, "Checking messages for validity. So far looking good.");
+
+    if (check_done_)
+    {
+      return;
+    }
+
+    msgs_.push_back(scan);
+
+    // To have only one call on the promise the subscriber is shut down
+    if (msgs_.size() == n_msgs)
+    {
```

Comment:
add atomic integer (unsigned long) for counting comparisons -> prevent success in case no comparisons were performed.